File: Data Structures & Algorithms/trapping-rain-water/submission-0.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        prefix, suffix = [0] * n, [0] * n
        res = 0

        # use monotonic stack
        st = []

        # first build the prefix array
        for i in range(n):
            while st and st[-1] < height[i]:
                st.pop()
            
            if not st:
                prefix[i] = height[i]
            else:
                prefix[i] = st[0]
            
            st.append(height[i])
        
        st.clear()

        # then buld the suffix array
        for i in range(n - 1, -1, -1):
            while st and st[-1] < height[i]:
                st.pop()
            
            if not st:
                suffix[i] = height[i]
            else:
                suffix[i] = st[0]

            st.append(height[i])

        for i in range(n):
            res += min(prefix[i], suffix[i]) - height[i]
        
        return res

File: Data Structures & Algorithms/validate-parentheses/submission-0.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<int> st;
        map<char, int> mp = {
            {'(', 1},
            {')', 2},
            {'{', 3},
            {'}', 4},
            {'[', 5},
            {']', 6}
        };

        for (int i = 0; i < s.size(); i++) {
            int idx = mp[s[i]];
            if (idx % 2 == 0) {
                if (st.empty())
                    return false;
                else if (mp[st.top()] != idx - 1)
                    return false;
                else
                    st.pop();
            } else {
                st.push(s[i]);
            }
        }

        return st.empty();
    }
};
